### PR TITLE
Use subcontexts directly instead of free functions

### PR DESCRIPTION
**Pull request by aleokdev**
Deprecates the old free functions and uses the subcontexts directly, as mentioned in #1012 and discussed in the Matrix room. WIP.

Note: The graphical context hasn't been refactored since the changes would be overwritten by the WGPU PR anyways.

### CODE REVIEW

**Comment by aleokdev** [comment]
Completely forgot about Filesystem, sorry.

File: src/audio.rs [hunk 1]
```diff
@@ -17,29 +18,17 @@ use std::sync::Arc;
...
 /// of your `Context` object.
-pub(crate) struct RodioAudioContext {
+pub struct AudioContext {
     _stream: rodio::OutputStream,
     stream_handle: rodio::OutputStreamHandle,
 }
 
-impl RodioAudioContext {
+impl AudioContext {
     /// Create new `RodioAudioContext`.
```
Comment (PSteinhaus) [inline src/audio.rs:32]: /// Create new `AudioContext`.
(or something similar)

File: src/context.rs [hunk 2]
```diff
@@ -35,20 +35,21 @@ use crate::timer;
...
+    pub time: timer::TimeContext,
+    /// Audio context.
+    #[cfg(feature = "audio")]
+    pub audio: audio::AudioContext,
+    /// Keyboard input context.
+    pub keyboard: input::keyboard::KeyboardContext,
+    /// Mouse input context.
+    pub mouse: input::mouse::MouseContext,
+    /// Gamepad input context.
+    pub gamepad: input::gamepad::GamepadContext,
```
Comment (PSteinhaus) [inline src/context.rs:52]: This needs the same treatment as audio, as it only compiles as long as `#[cfg(feature = "gamepad")]` is true.
```suggestion
    /// Gamepad input context.
    #[cfg(feature = "gamepad")]
    pub gamepad: input::gamepad::GamepadContext,
```

File: Cargo.toml [hunk 3]
```diff
@@ -31,11 +31,13 @@ name = "ggez"
 path = "src/lib.rs"
 
 [features]
-default = ["c_dependencies"]
+default = ["c_dependencies", "audio", "gamepad"]
 bzip2 = ["zip/bzip2"]
 mp3 = ["rodio/mp3"]
 multithread-image-decoding = ["image/hdr", "image/jpeg_rayon"]
 c_dependencies = ["bzip2", "mp3"]
```
Comment (PSteinhaus) [inline Cargo.toml:38]: Currently opting out of rodio by disabling "audio" would still fail, as "mp3" still links to rodio and is part of the default features through "c_dependencies", leading to rodio being loaded and compiled anyway.

**Review by PSteinhaus — CHANGES_REQUESTED** [review]
Overall, it looks good. The central problematic is that the ripple effects of disabling either of the new features haven't been taken care yet. So there needs to be a little (in the case of audio), or even a bit more (in the case of gilrs) conditional compilation to make things work. Luckily, these passages are easy to find by disabling the two features (temporarily taking them out of the default feature list for example) and then bumping your head until everything compiles.

File: Cargo.toml (hunk 3 above)
Comment (aleokdev) [reply Cargo.toml:38]: What's `c_dependencies` for?

Comment (aleokdev) [reply Cargo.toml:38]: > Currently opting out of rodio by disabling "audio" would still fail

Also, this is not possible; You can't disable a single feature. You need to disable all default features to enable one specifically.

Comment (PSteinhaus) [reply Cargo.toml:38]: `c_dependencies` groups together all libraries that at some point depend on a C-compiler, which can be undesirable, and luckily what they provide is all non-essential so they may be disabled.

You’re right, it’s a silly thought. Who would disable default features and then reenable `c_dependencies` without `audio`? And if you only need `bzip` or `mp3` but not the other there’s already features for that so it’s fine.

**Comment by PSteinhaus** [comment]
Meh, mistook `RodioAudioContext` for `RodioAudioDevice` in the commit description...
I have a feeling that more similar minor comment issues are still hidden in this, but these would be really minor, as the general sense behind them should remain logical, so I'm good with merging this now.

**Review by PSteinhaus — APPROVED** [review]

**Review by nobbele — APPROVED** [review]

**Pull request merged**